src/GraphBLAS-sharp/Implementations.fs

```diff
@@ -115,26 +116,549 @@ type CSRMatrix<'a when 'a : struct and 'a : equality>(csrTuples: CSRFormat<'a>)
+
+                    if boundaryX < 0 || boundaryY >= 0 && (firstRow < secondRow || firstRow = secondRow && firstColumn < secondColumn)
+                    then
+                        allRowsBuffer.[i] <- secondRowsBuffer.[boundaryY]
+                        allColumnsBuffer.[i] <- secondColumnsBuffer.[boundaryY]
+                        allValuesBuffer.[i] <- secondValuesBuffer.[boundaryY]
+                    else
+                        allRowsBuffer.[i] <- firstRowsBuffer.[boundaryX]
+                        allColumnsBuffer.[i] <- firstColumnsBuffer.[boundaryX]
+                        allValuesBuffer.[i] <- firstValuesBuffer.[boundaryX]
+            @>
+
+        let createSortedConcatenation =
+            opencl {
+                let binder kernelP =
+                    let ndRange = _1D(allRows.Length)
+                    kernelP
+                        ndRange
+                        firstRows
+                        firstColumns
+                        firstValues
+                        secondRows
+                        secondColumns
+                        secondValues
+                        allRows
+                        allColumns
+                        allValues
+                do! RunCommand createSortedConcatenation binder
+            }
+
+        let auxiliaryArray = Array.init allRows.Length (fun _ -> 1)
+
+        let fillAuxiliaryArray =
+            <@
+                fun (ndRange: _1D)
+                    (allRowsBuffer: int[])
+                    (allColumnsBuffer: int[])
+                    (allValuesBuffer: 'a[])
+                    (auxiliaryArrayBuffer: int[]) ->
+
+                    let i = ndRange.GlobalID0
+
+                    if allRowsBuffer.[i] = allRowsBuffer.[i + 1] && allColumnsBuffer.[i] = allColumnsBuffer.[i + 1]
+                    then
+                        auxiliaryArrayBuffer.[i + 1] <- 0
+                        allValuesBuffer.[i] <- (%plus) allValuesBuffer.[i] allValuesBuffer.[i + 1]
+            @>
+
+        let fillAuxiliaryArray =
+            opencl {
+                let binder kernelP =
+                    let ndRange = _1D(allRows.Length - 1)
+                    kernelP
+                        ndRange
+                        allRows
+                        allColumns
+                        allValues
+                        auxiliaryArray
+                do! RunCommand fillAuxiliaryArray binder
+            }
+
+        let dropExplicitZeroes =
+            <@
+                fun (ndRange: _1D)
+                    (allValuesBuffer: 'a[])
+                    (auxiliaryArrayBuffer: int[]) ->
+
+                    let i = ndRange.GlobalID0
+
+                    if allValuesBuffer.[i] = zero
+                    then auxiliaryArrayBuffer.[i] <- 0
+            @>
+
+        let dropExplicitZeroes =
+            opencl {
+                let binder kernelP =
```

Comment:
Should ```binder``` be declared inside ```opencl```?

Reply:
There are a lot of binders and I decided to avoid making up unique names for each one